Data/Primitive/Types.hs

```diff
@@ -208,6 +208,8 @@ derivePrim(FunPtr a, FunPtr, sIZEOF_PTR, aLIGNMENT_PTR,
            indexAddrArray#, readAddrArray#, writeAddrArray#, setAddrArray#,
            indexAddrOffAddr#, readAddrOffAddr#, writeAddrOffAddr#, setAddrOffAddr#)
 
+-- This is a workaround, since GeneralizedNewtypeDeriving doesn't work for Prim.
```

Comment:
Why doesn't `GeneralizedNewtypeDeriving` work for `Prim`, out of curiosity?

Reply:
@RyanGlScott Don't know either, it was giving a compile error.

Reply:
I ask because this code works for me:

```haskell
{-# LANGUAGE CPP #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE UnboxedTuples #-}

#if __GLASGOW_HASKELL__ >= 800
{-# LANGUAGE TypeInType #-}
#endif
module Bug where

import GHC.Exts

class Prim a where
  readByteArray# :: MutableByteArray# s -> Int# -> State# s -> (# State# s, a #)

instance Prim Char

newtype MyChar = MyChar Char
  deriving Prim
```

So why wouldn't it work here?

Reply:
Right. But that's my point—the error message tells you to enable `TypeInType`. So why not do so, as in the code in https://github.com/haskell/primitive/pull/123#discussion_r178567461?

Reply:
@RyanGlScott, we can only use `TypeInType` with new GHC versions, so I don't think that's a viable option.

Reply:
@RyanGlScott I'm very skeptical about introducing an extension like `TypeInType` in such a trivial use case. But seems you're not concerned so using that anyway.

Reply:
I disagree. Let me give some context on this error message.

You'll only ever experience this with GHC 8.4.1 or later, since that's the first version that featured a fix for https://ghc.haskell.org/trac/ghc/ticket/14579. The content of the #14579 isn't terribly important, except for the fact that it fixed a bug in `GeneralizedNewtypeDeriving` by explicitly inserting kinds in places that might need them.

A consequence of this change is that if you're GND'ing a method that involves unboxed tuples, such as `readByteArray#`, then you'll generate code that has explicit `TYPE (TupleRep ...)` kinds. This requires the use of `TypeInType`, hence the error message.

Thus, we only need to enable `TypeInType` on newer GHC versions, since it's a problem that's unique to newer GHC versions. (Technically, we could just as well do `#if __GLASGOW_HASKELL__ >= 804`, but I picked `800` since that's the version GHC version that actually sported `TypeInType`.)

Reply:
Thanks for the clarification. GHC really is an adventure :)